### ISSUE TITLE
[Discussion] Create our own notification system pop-up

### ISSUE THREAD

**Issue by RowenStipe** [qTox/qTox#2921]
In my research into trying to make heads or tails of https://github.com/tux3/qTox/pull/2771

To make it work properly on my own system, let alone others, I've come to the conclusion that "Native" Notifications would be very frustrating; and much alcohol would be consumed.
So since notification popups are something that users are interested in having. I'm proposing we attempt to accomplish it by creating our own, which I'm sure a few people would rather go the native route.

My idea is to basically use a borderless window in Qt 5 to display the notification pop-up.
Now with this idea comes the option to also design the notification box to be more pleasing to look at, in some aspect anyways. Notifications could also act for input for special events such as a file transfer or incoming call, along with message notifications. And creates the option of letting the user set where the notification box appears.

Going the native route would certainly look cleaner for certain systems. Not all systems have a notification backend, and not all notification systems are built the same. (obviously ... ) I just think that it's unnecessary to try and program for every notification back end than provide our own. 

Anyways, here's some mock ups from sender Tux :

![qtox-notifications-mock-new-message](https://cloud.githubusercontent.com/assets/1356364/12774214/bbe21404-ca10-11e5-9fde-26f1c6d8875a.png)
The black border could change depending on the theme the user has set.

![qtox-notifications-mock-file-transfer](https://cloud.githubusercontent.com/assets/1356364/12774252/0c38c34e-ca11-11e5-91e0-03ddf112f772.png)
Icon could change depending on file type

![qtox-notifications-mock-incoming-call](https://cloud.githubusercontent.com/assets/1356364/12774269/2c54db86-ca11-11e5-86c4-9356afa90571.png)
This one just seemed obvious to have.

Anyways, thoughts? Felt like this was something that might need some discussion.
## End credits

Example User used: @tux3  
Fonts:
- DejaVu Sans
- Font Awesome

Dimensions:
- Notification Box: 400x100 px
- Theme color box: 100x100 px
- Avatar in notification: 80x80 px

Lorem ipsum message:
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed sit amet mauris pellentes(...)que ante mollis hendrerit. Aliquam erat volutpat. Curabitur lorem mauris, bibendum at enim.

**Comment by agilob** [comment]
I made a working prototype in like 3hours... I also used borderless window + transparency, some days later my device was stolen https://github.com/tux3/qTox/issues/483#issuecomment-59916964

**Comment by RowenStipe** [comment]
... Wow That sucks major balls. =(

**Comment by agilob** [comment]
I can try to find a backup somewhere on old drives, but it was really 2-3hours to create a prototype. I can do it again, the bigger problem was testing on different envs, OXS, Windows, Gnome, kde4/5... and inserting notification system into existing codebase. Adding it to dozens of actions like group invite, friend invite, call etc. If you want to take care of it will create a branch this week with that borderless windows + transparency etc, but as said it was the easier part.

**Comment by RowenStipe** [comment]
Well, if you don't mind working out the basic borderless window prototype, I certainly wouldn't mind figuring out inserting it into the current code base and setting it up to actually function.

**Comment by anoadragon453** [comment]
I can test Windows, OSX, Gnome, XFCE.

**Comment by agilob** [comment]
Can't promise because I have others projects going on right now, but I can start it on Saturday/Sunday, I'll provide a link here later.

**Comment by ghost** [comment]
Count me in for Cinnamon (and KDE, but that's most likely covered).

Adapting qBittorrent's code is not an option?

**Comment by agilob** [comment]
Does qbittorent has notification? oO Never seen any. Might give it a try, qBT has very good quality code, but we need notifications with buttons (Yes/No, Accept/Cancel, Answer/Reject).

I have KDE5 on arch and xfce4 on FBSD ;) afair Cinammon had a lot of problems handling transparent Qt windows.

**Comment by RowenStipe** [comment]
After doing a big of digging, I've come to the conclusion that core.cpp is the perfect file to place hooks for a notification back end since it deals with everything that would need a notification.

**Comment by bhelm** [comment]
using custom notifications instead of native ones is a good idea, especially because of the flexibility this offers.

i can imagine situations when native notification they are still nice because they are part of a larger ecosystem. i guess there are plugins that can do nice things with system notifications.
the linux skype allows you to enable both skypes custom and native notifications at once. i have both enabled, because they are displayed on different screens of my 3 screen setup. 
if i look at screen 1, i do not notice anything on screen 3.

**Comment by LuccoJ** [comment]
I can imagine that in _most_ situations it's nicer to use system notifications... there is a reason notifications are almost universally a part of the core environment, and often tweaked and improved by projects like GNOME and KDE.

Actually, have a look at GNOME's notification system when used by Empathy: it actually lets you _reply to messages from inside the notification!_

**Comment by anoadragon453** [comment]
@LuccoJ No reason we couldn't have quick-reply functionality within our custom notifications.

**Comment by agilob** [comment]
There is also no reason why use we use custom, own theme, when we could just use standard system default skin.

**Comment by ProMcTagonist** [comment]
The reason is outlined in the OP. It's a pain to integrate all the different platform-appropriate notification systems.

**Comment by DeerSpotter** [comment]
i would love to test windows notifications as soon as possible, count me in. I have 14 computers running windows 10 with the client installed.

**Comment by anoadragon453** [comment]
Well our own notification system would probably be both faster/easier to get working and potentially have more functionality, so I say we start with that method.

**Comment by agilob** [comment]
https://gitlab.com/agilob/qfiction

Please test if transparency and position of widget is correct, it should be like that:
![screenshot_20160206_154338](https://cloud.githubusercontent.com/assets/2544251/12867556/6c42d6ec-cce8-11e5-9c88-eee3c0957c1a.png)

So far it's mostly UI mockup, no functionality, will do some more work later today and tomorrow.

**Comment by ghost** [comment]
I wanted to pitch in since I'm also interested in this feature. My initial preference would of course be native notifications, but if there are no wrapper libraries that handle all the specific back ends for you, I understand that this can be a pain to maintain. Would it perhaps be an option to somehow offer best of  both worlds? This would include not supporting specific back ends, but exposing some kind of service that allows users to do this for you. An example of this is Pidgin: it has its own "notifications" (if you can call it that) but a user has created a GNOME shell extension that can interact with Pidgin and display native notifications instead. Another example is the Atom editor: it has its own notifications for various error messages but you can install a package that displays your OS' native notification instead. This way part of the load is taken off your shoulders (you could optionally provide some common back ends if you're up to it) but you don't prevent anyone from making it possible outside the application itself.

**Comment by ghost** [comment]
> qfiction

The notifications are in the right screen but the wrong place (upper right corner instead of bottom right, too high) and they don't disappear unless I finish qfiction.

Before any notification

![01](https://cloud.githubusercontent.com/assets/6382342/12869008/e295d396-cd13-11e5-8409-3f15ed94bf24.png)

After 1 notification

![02](https://cloud.githubusercontent.com/assets/6382342/12869011/eb5c8bbe-cd13-11e5-95c6-df6e3664d3f1.png)

After 2+ notifications

![03](https://cloud.githubusercontent.com/assets/6382342/12869013/eefcea70-cd13-11e5-9d58-3b86383a6b8b.png)

---

`clang` -> is that dependency temporary? Right now qTox doesn't need it and is huge.

---

The notifications are always displayed in this position

> $ ./qfiction 
> x:  1544   y:  61

---

Cinnamon 2.8.6
Qt 5.5.1

**Comment by agilob** [comment]
> the right screen but the wrong place (upper right corner instead of bottom right)

That's correct, it should be in right top corner, bottom right is wrong and you should feel bad.

> the display is broken and they don't disappear unless I finish qfiction.

Which is what should happen, I just asked to test if you have transparency and experience bug with multiscreen setup.

I don't understand anything else from your comment.

**Comment by ghost** [comment]
> That's correct, it should be in right top corner

Eww, no, nothing goes where I change tabs.

> I don't understand anything else from your comment.

The images are how I see it (ie cropped), `clang` is the extra dependency I had to install to compile, `x: 1544 y: 61` is qfiction's output in my terminal. Post edited.

**Comment by anoadragon453** [comment]
@kehugter @agilob Guys guys. 

Why not let the user select?

![tmp_26189-why_not_both637804024](https://cloud.githubusercontent.com/assets/1342360/12869127/e24bd3f2-ccca-11e5-9712-6abc5ace508a.jpg)

**Comment by agilob** [comment]
My CI (FBSD) doesn't have gcc/g++ I didn't bother installing it, so it is clang, it will work with g++ just [remove two lines from .pro file](https://gitlab.com/agilob/qfiction/commit/a7f3af18e675bb968ca12f61c0d0a8d3daf682ab).

> x: 1544 y: 61 

That's location on the screen where centre of the notification is, meaningless for you unless you have 3 screens, then those values would tell me how wide/high screens setup it. Notification should appear on the main screen.

Notification appear one on another, I'll fix it so coder will provide a number of notification and next one will appear below n-1. This is just to test transparency and multiscreen setup as I see those issues are priority over "interactivity" and style of the widget.

**Comment by anoadragon453** [comment]
@kehugter @agilob Guys, guys.

Why not have the user pick the location via dropdown?

![why_not_both](https://cloud.githubusercontent.com/assets/1342360/12870088/4692fb56-cce9-11e5-8b54-3ff4bf0d1155.jpg)

**Comment by RowenStipe** [comment]
At this point we may as well have a separate tab in settings to deal with notifications.

**Comment by RowenStipe** [comment]
![OSX Screenshot](https://cloud.githubusercontent.com/assets/1356364/12870991/c7af81ce-cd2e-11e5-8b5c-46fd27ce76dd.png)

How it looks on OSX, I have a dual screen set up and it's appearing to the farthest right in the screen set up.

**Comment by RowenStipe** [comment]
Also update on my own progress, I've got basic debug outputs setup for [new messages](https://gist.github.com/RowenStipe/ab652ce2485159b8de91) and I have friend online status notifications working but it has lead me to [some interesting output](https://gist.github.com/RowenStipe/4cdbe303aa1464dc1a8f)

**Comment by anoadragon453** [comment]
Notification placement looks fine on i3, but the notification never disappears, or has any content. Single monitor, 2880x1620.

![ss](https://cloud.githubusercontent.com/assets/1342360/12875558/cb6018dc-cda2-11e5-9f0a-2068ce733ff7.png)

On commit 0d958b68

Log:

```
Starting ~/Documents/code/cpp/build-qfiction-Desktop-Debug/qfiction...
x:  2504   y:  61
x:  2504   y:  61
x:  2504   y:  61
x:  2504   y:  61
x:  2504   y:  61
x:  2504   y:  61
```

Edit: I should also note that there is no transparency with either one or more notifications.

**Comment by anoadragon453** [comment]
Looks good on XFCE4, however notification continues to remain indefinitely. One monitor, different resolution of 2560x1440.

![screenshot_2016-02-07_14-12-24](https://cloud.githubusercontent.com/assets/1342360/12875716/f91ed572-cda4-11e5-85a3-c10cc33d86b1.png)

```
x:  2184   y:  61
```

Same commit 0d958b68

**Comment by anoadragon453** [comment]
Here it is on GNOME3 with scaling at 2x for HiDPI. Elements appear but are squished together. Remains indefinitely. Single monitor, 2880x1620.

![screenshot from 2016-02-07 14-16-48](https://cloud.githubusercontent.com/assets/1342360/12875766/e5d4ee10-cda5-11e5-9387-9f35b9eff957.png)

```
x:  2504   y:  61
```

Same commit 0d958b68

**Comment by agilob** [comment]
I made an update with the widget, not it disappears after n seconds, no need to test now.

I'll ask you  @anoadragon453 to test styling because you have some wired styling manager, it shouldn't be that big.

**Comment by anoadragon453** [comment]
@agilob Sounds good, let me know when such changes are made.

**Comment by agilob** [comment]
Can I reproduce your setup using something like 144DPI or `export QT_DEVICE_PIXEL_RATIO=2`? What's `2x for HiDP` anyway?

**Comment by anoadragon453** [comment]
You can try QT's Device Ratio. My DPI is apparently 96x96 as reported by `xdpyinfo`, however it all really stems from setting `Window Scaling` to 2 in Gnome Tweak Tool.

**Comment by agilob** [comment]
An update for you:
The widget is more interactive now, decays, disappears, emits signals when ok/no was clicked. The next stage is positioning in on "stack-like" way if more than one should be shown.

**Comment by anoadragon453** [comment]
Widget's Accept and Reject buttons work great, they fire every time.

In terms of looks you can actually make out some faint outline button boxes on i3 now:

![ss2](https://cloud.githubusercontent.com/assets/1342360/12939863/e49722be-cf78-11e5-8178-1c254e4c83f5.png)

But the text is still invisible. Perhaps it's a font issue? i3 is pretty bare bones...

Looks great on XFCE4 on 2560x1440 still

Still looks cramped on GNOME3 with 2880x1620:

![screenshot from 2016-02-09 21-59-10](https://cloud.githubusercontent.com/assets/1342360/12939899/278cb43a-cf79-11e5-96b1-4a93d313a76e.png)

I think the text is the right size, just the black box needs to be a bit bigger. This is a 15in laptop I'm on.

[Output of xrandr](https://gist.github.com/anoadragon453/1fb81781ef19df37c487) if it's any use.

**Comment by agilob** [comment]
There's no styling yet https://gitlab.com/agilob/qfiction/issues

**Comment by anoadragon453** [comment]
Ah, I misunderstood:

> An update for you:

Consider it free testing? :P

**Comment by agilob** [comment]
> Consider it free testing? :P

Yes, "test if you bother" ;)

**Comment by agilob** [comment]
![screenshot_20160210_174230](https://cloud.githubusercontent.com/assets/2544251/12956168/81bb49da-d01e-11e5-8e97-a644fcfd367b.png)

After making screenshot I decreased spacing between stacked widgets, from 20 to 10.

**Comment by anoadragon453** [comment]
Looks pretty good for the most part. I can still find myself clicking the button without a new notification appearing however. This is most notable when half of the notifications have disappeared, with the other half still to go. Pressing "Show Notification" doesn't seem to add a new notification anywhere. What I think it happening is that it's putting it under the newest notification, and we can't see any change because it's overlapping it.

Most chat apps I know start from index 0 again as soon as it's free and so on, do you think we can try emulating this behaviour? Otherwise for the most part the spacing and everything else is good.

**Comment by agilob** [comment]
> Most chat apps I know start from index 0 again as soon as it's free and so on, do you think we can try emulating this behaviour? Otherwise for the most part the spacing and everything else is good.

Yes, but it's not a problem with notification but how they're handled. It will be implemented by @RowenStipe.

**Comment by anoadragon453** [comment]
Any progress on this @RowenStipe @agilob ?

**Comment by agilob** [comment]
Yea, sorry for ignoring this topic for so long, at first I was busy with other project, then I forgot about it, sorry :<

The notification widget is nearly ready, anyone can start attaching it to qTox, the task was really simple. There are some places where I need to polish it a bit, but I think it's in usable state right now. I will experiment with different styles, one light (qtox based) and one dark for the widget. Everyone is welcome to test it.

![screenshot_20160426_205002](https://cloud.githubusercontent.com/assets/2544251/14832115/78ad0b22-0bf0-11e6-9c74-12c1988729b0.png)

**Comment by SkyzohKey** [comment]
Why not simply using libnotify ?

**Comment by anoadragon453** [comment]
@SkyzohKey Does libnotify work on Windows?

Edit: Also more dependencies are always a con.

**Comment by SkyzohKey** [comment]
@anoadragon453 

> Does libnotify work on Windows?

Nope, but you could use libnotify/gnome notification on linux since they work on every DE and simply use your own on Windows. + Growl on OSX. Would probably avoid many bloat

> Also more dependencies are always a con.

`libnotify` package exists in **every** distro ^^

**Comment by anoadragon453** [comment]
@SkyzohKey Using three notification backends for all OS' sounds like more bloat than what @agilob is proposing with a Qt widget instead, however the question is integrating it into each DE. 

If it were up to me, I'd say get a working implementation first that might not look great, but gets the job done, then in future releases we can integrate into the rest of the OS.

**Comment by SkyzohKey** [comment]
@anoadragon453 So you prefer to reinvent the wheel on 3 OS instead of only one (Windoesn't) ?

Linux has libnotify/GLib.Notification/etc,
OSX has Growl,
Windows has... nothing, like ever.

IMHO it would be more intelligent to only focus on:
1. Integrating desktop notifications on Linux and OSX using proved libraries that integrates well in every distro/window manager.
2. Developing your own notification system for Windows, or port libnotify to it.

**Comment by antis81** [comment]
@SkyzohKey Your second point goes totally contrary to what you wrote before. Implementing a notification API on Wndows means to implement an API, that (maybe with minor differences) works on any Platform. AFAIK Growl is supported on Windows btw, but has to be installed separately. I didn't touch that topic for years though.

If multiplatform is a point here, than I'd say to check, if a notification API exists (Growl or libnotify) and send message to it. Otherwise create a popup widget (as ugly as possible! :smile:).

**Comment by ghost** [comment]
Is [snorenotify](https://github.com/KDE/snorenotify) not an option? It has everything we need, based on Qt, backed by KDE and is future proof with Android support.

**Comment by antis81** [comment]
@kehugter It is an option (for mac/linux at least) -> obviously there's currently nobody interested in writing the code. Probably you want to write it and are prepared to "go through the deep valley of code review"… :smile:

**Comment by antis81** [comment]
@RowenStipe I currently am working on other important part on qTox and cannot bring this one to stable status. Do you think you will have the time and revive/cleanup this one?

**Comment by ghost** [comment]
@antis81 I'd be glad to, but it's way beyond my knowledge.

[Telegram's desktop client](https://github.com/telegramdesktop/tdesktop) has a nice Qt non-interactive notification system and a lengthy discussion about (Linux native) notifications too https://github.com/telegramdesktop/tdesktop/issues/91.

**Comment by WhyNotHugo** [comment]
Sorry to sound negative, but I must point out, building your your notifications sounds like a very very bad idea. It basically spits on the users choice of notification settings, completely overriding it (and looks/behaves alien). This includes:
- Inconsistent "away" support: Will you respect how me notification daemon handles notifications when I'm AFK? (probably not, because you've no programmatic way of knowing it).
- Colour, theme, _position_ (lots of re-implementations chose positions that annoy users by covering up critical screen areas, like the).
- Inconsistent key-bindings (Will hide/"bring back" key-bindings be the same as other notifications? What about other interaction key-bindings?)
- Adding to the last point, what about accessibility, including actions in them, etc (remember, these pop-ups won't have focus).
- How do I temporarily disable notifications? Do I need to disable my system ones _and then_ qTox as well, manually (image if I had to do this per-app!).
- Wheel reinvention: There's dozens of other quirks (as well as some mentioned above) regarding notification daemons, which each implementation handles differently. Each user has picked one. By using another, you're re-introducing those quirks for each user.

**Comment by zetok** [comment]
@hobarrera don't worry about sounding negative, as long as there's valid criticism going on :)

So, most of your points could be used to criticize current notification daemons, not just a custom one. Frankly, most of system notification daemons are shit, assuming that they're available at all. And no, users don't actually pick them, they use whatever their shitty DE provides.

Considering that notifications in qTox have to work & provide a minimal set of features, no, relying only on broken notification daemon simply won't cut it.

About accessibility – I really wish that qTox could be better with this, but then again, AFAIK it doesn't support it at all (or for some reason I wasn't able to get accessibility working for Qt5 even though it worked for GTK stuff).

_sigh_

Hopefully someone could [help with it](https://github.com/qTox/qTox/milestone/1).

> How do I temporarily disable notifications? Do I need to disable my system ones and then qTox as well, manually (image if I had to do this per-app!).

Yes. Sorry, but the alternative is to have utterly broken notifications, which is a no-go.

# 

Anyway, there is PR #2771 which does implement support for several notification backends, but it needs some love & care. I wouldn't be against supporting multiple backends – the thing that I require from it is that **it has to work**.

It would be great if someone could take over the PR & get it into proper shape.

**Comment by WhyNotHugo** [comment]
> So, most of your points could be used to criticize current notification daemons, not just a custom one. 

Only accesibility/keybinding support seem to partially apply to some existing ones. The rest isn't.

That aside though, I feel like introducing _two_ notifications implementations into the users' systems, which behave differently, and have to be controlled separately, is even worse.
None of the points mentioned above are addresses. Maybe time would be better spent fixing your current daemon to fix whatever is failing on it. qTox would directly benefit, _and_ you'd have a fixed an extra problem.

My first three points are non-applicable for existing daemons. The following too only apply to _some_. The final one is I think what weighs the most and what will most likely tend to drive users away.
(FWIW, dunst is actually an implementation that solves _all_ the above mentioned issues).

**Comment by anthonybilinski** [comment]
We now have custom notifications implemented by snorenotify. It will be available as a compile option in v1.17.0, or current master tip.

Their use can be expanded and dep compilation needs to be added for windows, but I think snorenotify is the way we'll move forward for notifications, so closing this issue.
